pkg/worker/verifier/transactional_producer_worker.go

```diff
@@ -0,0 +1,345 @@
+		fmt.Fprintf(&key, "ABORTED MSG: %06d.%018d", producerId, sequence)
+	}
+
+	payload := make([]byte, pw.config.messageSize)
+
+	var r *kgo.Record = kgo.KeySliceRecord(key.Bytes(), payload)
+
+	if pw.fakeTimestampMs != -1 {
+		r.Timestamp = time.Unix(0, pw.fakeTimestampMs*1000000)
+		pw.fakeTimestampMs += 1
+	}
+	return r
+}
+
+type TransactionalProducerWorkerStatus struct {
+	// How many messages did we try to transmit?
+	Sent int64 `json:"sent"`
+
+	// How many messages did we send successfully (were acked
+	// by the server at the offset we expected)?
+	Acked int64 `json:"acked"`
+
+	// How many messages landed at an unexpected offset?
+	// (indicates retries/resends)
+	BadOffsets int64 `json:"bad_offsets"`
+
+	// How many failures occured while trying to begin, abort,
+	// or commit a transaction.
+	FailedTransactions int64 `json:"failed_transactions"`
+
+	// How many times did we restart the producer loop?
+	Restarts int64 `json:"restarts"`
+
+	// Ack latency: a private histogram for the data,
+	// and a public summary for JSON output
+	latency metrics.Histogram
+	Latency worker.HistogramSummary `json:"latency"`
+
+	Active bool `json:"latency"`
+
+	lock sync.Mutex
+
+	// For emitting checkpoints on time intervals
+	lastCheckpoint time.Time
+}
+
+func NewTransactionalProducerWorkerStatus() TransactionalProducerWorkerStatus {
+	return TransactionalProducerWorkerStatus{
+		lastCheckpoint: time.Now(),
+		latency:        metrics.NewHistogram(metrics.NewExpDecaySample(1024, 0.015)),
+	}
+}
+
+func (self *TransactionalProducerWorkerStatus) OnAcked() {
+	self.lock.Lock()
+	defer self.lock.Unlock()
+	self.Acked += 1
+}
+
+func (self *TransactionalProducerWorkerStatus) OnBadOffset() {
+	self.lock.Lock()
+	defer self.lock.Unlock()
+	self.BadOffsets += 1
+}
+
+func (pw *TransactionalProducerWorker) produceCheckpoint() {
+	err := pw.validOffsets.Store()
+	util.Chk(err, "Error writing offset map: %v", err)
+
+	data, err := json.Marshal(pw.Status)
+	util.Chk(err, "Status serialization error")
+	log.Infof("TransactionalProducer status: %s", data)
+}
+
+func (pw *TransactionalProducerWorker) Wait() error {
+	pw.Status.Active = true
+	defer func() { pw.Status.Active = false }()
+
+	n := int64(pw.config.messageCount)
+
+	for {
+		n_produced, bad_offsets, err := pw.produceInner(n)
+		if err != nil {
+			return err
+		}
+		n = n - n_produced
+
+		if len(bad_offsets) > 0 {
+			log.Infof("Produce stopped early, %d still to do", n)
+		}
+
+		if n <= 0 {
+			return nil
+		} else {
+			// Record that we took another run at produceInner
+			pw.Status.Restarts += 1
+		}
+	}
+}
+
+func (pw *TransactionalProducerWorker) produceInner(n int64) (int64, []BadOffset, error) {
+	opts := pw.config.workerCfg.MakeKgoOpts()
+	randId := uuid.New()
+
+	opts = append(opts, []kgo.Opt{
+		kgo.ProducerBatchCompression(kgo.NoCompression()),
+		kgo.RequiredAcks(kgo.AllISRAcks()),
+		kgo.RecordPartitioner(kgo.ManualPartitioner()),
+		kgo.TransactionalID("p" + randId.String()),
```

Comment:
nit: mind including hostname too in the ID for tracing back to a verifier?  (p-docker-123-xyz)